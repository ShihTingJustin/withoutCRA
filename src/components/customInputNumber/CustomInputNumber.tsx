import React, {
  useEffect,
  useState,
  useRef,
  useImperativeHandle,
  forwardRef,
  Ref,
  InputHTMLAttributes,
  FocusEventHandler,
  FocusEvent,
  ChangeEvent
} from 'react';
import { AiOutlinePlus, AiOutlineMinus } from 'react-icons/ai';
import Button from '../button/Button';

import './customInputNumber.scss';

type ValueType = string | number;
interface InputNumberProps<T extends ValueType = ValueType>
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'value' | 'onChange'> {
  min?: number;
  max?: number;
  step?: number;
  name?: string;
  value?: number;
  disabled?: boolean;
  className?: string;
  minusDisabled?: boolean;
  plusDisabled?: boolean;
  yetDistributedCount?: number;
  onBlur?: FocusEventHandler<HTMLSpanElement>;
  onChange?: (e: ChangeEvent<HTMLInputElement>) => void;
}

const CustomInputNumber = forwardRef(
  (props: InputNumberProps, ref: Ref<HTMLInputElement | null>) => {
    const {
      min = 0,
      max = 10,
      step = 1,
      value = 0,
      disabled = false,
      minusDisabled = false,
      plusDisabled = false,
      yetDistributedCount = 0,
      onChange,
      onBlur,
      ...inputProps
    } = props;

    const inputRef = useRef<HTMLInputElement | null>(null);
    useImperativeHandle(ref, () => inputRef?.current);

    const [stateValue, setValue] = useState<number>(Number(value));
    const [stateMinusDisabled, setMinusDisabled] = useState(minusDisabled);
    const [statePlusDisabled, setPlusDisabled] = useState(plusDisabled);

    const finalMinusDisabled = disabled || minusDisabled || stateMinusDisabled;
    const finalPlusDisabled = disabled || plusDisabled || statePlusDisabled;

    useEffect(() => {
      setMinusDisabled(stateValue <= min);
      setPlusDisabled(stateValue >= max);

      if (stateValue >= min && stateValue <= max) {
        handleInputEventManually();
      }
    }, [stateValue, yetDistributedCount]);

    const correctValue = () => {
      if (stateValue < min) setValue(min);
      if (stateValue > max) {
        if (yetDistributedCount === 0) {
          setValue(min);
        } else {
          setValue(max);
        }
      }
    };

    const handleBlur = (e: FocusEvent<HTMLInputElement>) => {
      correctValue();
      onBlur?.(e);
    };

    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
      if (disabled) return;
      const value = Number(e.target.value);
      const notNumber = Number.isNaN(value);
      if (notNumber) return e.preventDefault();
      setValue(value);
    };

    const handleInputEventManually = () => {
      correctValue();
      const nativeInputValueSetter = Object.getOwnPropertyDescriptor(
        window.HTMLInputElement.prototype,
        'value'
      )?.set;
      nativeInputValueSetter?.call(inputRef.current, String(stateValue));

      const inputEvent = new Event('input', { bubbles: true });
      inputRef.current?.dispatchEvent(inputEvent);
      onChange?.(inputEvent as unknown as ChangeEvent<HTMLInputElement>);
    };

    const handleMinus = () => {
      if (finalMinusDisabled) return;
      setValue((prev) => {
        const newValue = prev - step;
        const isValueValid = newValue >= min;
        return isValueValid ? newValue : prev;
      });
    };

    const handleAdd = () => {
      if (finalPlusDisabled) return;
      setValue((prev) => {
        const newValue = prev + step;
        const isValueValid = newValue <= max;
        return isValueValid ? newValue : prev;
      });
    };

    return (
      <div data-testid="customInputNumber" className="input-number-root">
        <Button disabled={finalMinusDisabled} onClick={handleMinus}>
          <span>
            <AiOutlineMinus size={30} color={disabled ? '#c0c0c0' : '#1e9fd2'} />
          </span>
        </Button>
        <input
          data-disabled={disabled}
          value={stateValue}
          ref={inputRef}
          disabled={disabled}
          onBlur={handleBlur}
          onChange={handleChange}
          {...inputProps}
        />
        <Button disabled={finalPlusDisabled} onClick={handleAdd}>
          <span>
            <AiOutlinePlus size={30} color={disabled ? '#c0c0c0' : '#1e9fd2'} />
          </span>
        </Button>
      </div>
    );
  }
);

export default CustomInputNumber;
