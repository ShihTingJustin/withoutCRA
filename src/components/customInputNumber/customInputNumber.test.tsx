import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen } from '@testing-library/react';

import CustomInputNumber from './CustomInputNumber';

describe('Test Button component', () => {
  const component = <CustomInputNumber  />;

  it('[Render] match snapshot', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('[Render] toBeInTheDocument', () => {
    render(component);
    const element = screen.getByTestId('customInputNumber');
    expect(element).toBeInTheDocument();
  });
});
