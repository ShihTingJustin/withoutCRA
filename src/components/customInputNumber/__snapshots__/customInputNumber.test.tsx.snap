// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button component [Render] match snapshot 1`] = `
<div
  className="input-number-root"
  data-testid="customInputNumber"
>
  <div
    className="button-root"
    data-testid="button"
  >
    <button
      data-disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
    >
      <span>
        <svg
          color="#1e9fd2"
          fill="currentColor"
          height={30}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": "#1e9fd2",
            }
          }
          viewBox="0 0 1024 1024"
          width={30}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M872 474H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h720c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8z"
          />
        </svg>
      </span>
    </button>
  </div>
  <input
    data-disabled={false}
    disabled={false}
    onBlur={[Function]}
    onChange={[Function]}
    value={0}
  />
  <div
    className="button-root"
    data-testid="button"
  >
    <button
      data-disabled={false}
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
    >
      <span>
        <svg
          color="#1e9fd2"
          fill="currentColor"
          height={30}
          stroke="currentColor"
          strokeWidth="0"
          style={
            {
              "color": "#1e9fd2",
            }
          }
          t="1551322312294"
          version="1.1"
          viewBox="0 0 1024 1024"
          width={30}
          xmlns="http://www.w3.org/2000/svg"
        >
          <defs />
          <path
            d="M474 152m8 0l60 0q8 0 8 8l0 704q0 8-8 8l-60 0q-8 0-8-8l0-704q0-8 8-8Z"
          />
          <path
            d="M168 474m8 0l672 0q8 0 8 8l0 60q0 8-8 8l-672 0q-8 0-8-8l0-60q0-8 8-8Z"
          />
        </svg>
      </span>
    </button>
  </div>
</div>
`;
